### DISCUSSION TITLE
KH "Powerful algorithm" is bit mysterious...

### DISCUSSION THREAD

**Discussion by jklymak** [CliMA/Oceananigans.jl#1283] [Ideas]
I was looking over the documentation for oceananigans.  It is looking very exciting, though I don't imagine I can seriously use it until MPI and immersed boundaries work.  

I found the discussion of the "Powerful algorithm" in https://clima.github.io/OceananigansDocumentation/stable/generated/kelvin_helmholtz_instability/ a bit more obscure than perhaps necessary.  It seems that you are successively filtering the solutions so that you emphasize the fastest growing mode, but keep the amplitude of the velocities in a regime that will let the model stay linear.  What I was missing is a clearer motivation of "why".  Can you include a couple of sentences about what the end goal of this would be, and why it is not achievable just running a free simulation?  I *think* its something like:

> The goal is to find the fastest growing linear mode, and have an idea of its growth rate to compare with theory.  This is difficult to extract empirically from a free simulation because the non-linearities quickly spread variance from the fastest growing mode to other scales.

but I really had to think to dredge that up, and I'm not sure if I am even correct.

**Comment by ali-ramadhan** [comment]
Hi @jklymak! I'll let @navidcy and @glwagner comment on the _powerful_ algorithm since they developed the example in #1125.

> I don't imagine I can seriously use it until MPI and immersed boundaries work.

Was just curious about your MPI concerns: how large are the simulations you'd like to run? Oceananigans can do 100~150 million grid points on a 32 GB GPU (e.g. a V100) which covers some simulations that tend to require multiple CPUs. Of course, if your simulations are larger or if large GPUs aren't super available then MPI + CPUs is the way to go.

@tomchor has also raised a similar concern about MPI (https://github.com/CliMA/Oceananigans.jl/issues/1234#issuecomment-738133452). MPI support is technically under-development (PR #590) but the priority keeps getting pushed...

I see you also came across @whitleyv's immersed boundary work (#1036). Hopefully in some months Oceananigans will have MPI and immersed boundary support!

Sorry to derail this thread! Happy to continue discussing in a new thread to keep this one on-topic.

**Reply by jklymak** [reply]
Not derailing at all.  I mostly posted here, rather than open an issue, because we are thinking of using GitHub discussion in matplotlib, but I hadn't seen it in action ;-)

Thats great that a GPU can do so much.  Recently a typical run might be something like typically run nz = 400, nx = 1000, ny = 300, so being able to do that on one GPU would be great (though it looks like [koehr](https://www.navydsrc.hpc.mil/docs/koehrUserGuide.html) the machine I currently have access to, only has 16 Gb on the GPU nodes), .    If I could get more resolution then I would do so.  The ability to harness GPUs is a huge attraction of Oceananigans. I imagine some of the LES folks have GPU enabled code, but their code tends to not be public.  

I study stratified flow over irregular topography, so I'd really need immersed boundaries.  OTOH, I'm not personally fussy about shaved cells etc.  I figure the ocean is rough, and if I don't like the stair-stepping, then I really need more resolution.  From a quick look at #1036 it looked like one of the concerns was whether the boundaries sit on the edge of the grid cells or not - seems that would be a good place to start.

**Reply by glwagner** [reply]
The problem size (400, 1000, 300) is close to the limit of what we can run on a single V100 with 32 GB of memory, give or take a few 10^6 DOF. But it seems likely that a problem of that size would benefit from multi-GPU parallelism too.

It seems that grid-coincident immersed boundaries (esp 2.5-D immersed boundaries) are dead simple. Arbitrary immersed boundaries are a bit more complicated. I agree that grid-coincident, stair-steppy immersed boundaries might be fine for a lot of applications but I'm not sure that's generally true. Bottom boundary layer problems, or turbulent boundary layer separation problems might need more accurate treatment of a smooth boundary... ?

**Reply by glwagner** [reply]
Looks like (400, 1000, 300) works:

```
glwagner@satori-login-002:~$ nvidia-smi
Thu Jan  7 09:59:05 2021       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 440.64.00    Driver Version: 440.64.00    CUDA Version: 10.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla V100-SXM2...  On   | 00000004:04:00.0 Off |                    0 |
| N/A   42C    P0    62W / 300W |  19347MiB / 32510MiB |      9%      Default |
+-------------------------------+----------------------+----------------------+
|   1  Tesla V100-SXM2...  On   | 00000035:03:00.0 Off |                    0 |
| N/A   40C    P0    38W / 300W |     10MiB / 32510MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0     51999      C   python3                                     9129MiB |
|    0     52103      C   python3                                     9129MiB |
|    0     75828      C   python3                                     1079MiB |
+-----------------------------------------------------------------------------+
glwagner@satori-login-002:~$ CUDA_VISIBLE_DEVICES=1 julia
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.4.1 (2020-04-14)
 _/ |\__'_|_|_|\__'_|  |  
|__/                   |

julia> using Oceananigans

julia> model=IncompressibleModel(grid=RegularCartesianGrid(size=(400, 1000, 300), extent=(1, 1, 1)), architecture=GPU(), tracers=:b, buoyancy=BuoyancyTracer())
IncompressibleModel{GPU, Float64}(time = 0 seconds, iteration = 0) 
├── grid: RegularCartesianGrid{Float64, Periodic, Periodic, Bounded}(Nx=400, Ny=1000, Nz=300)
├── tracers: (:b,)
├── closure: IsotropicDiffusivity{Float64,NamedTuple{(:b,),Tuple{Float64}}}
├── buoyancy: BuoyancyTracer
└── coriolis: Nothing
```

**Reply by glwagner** [reply]
I used just one tracer (buoyancy) to stack the deck in my favor 😏

**Reply by jklymak** [reply]
> Bottom boundary layer problems, or turbulent boundary layer separation problems might need more accurate treatment of a smooth boundary... ?

I think that really depends on the problem and application.  Do you anticipate oceananigans being used for precise DNS on small-scale flow over smooth bodies?  I admit shaved cells would be nice, but they come at a complexity cost, and in the vertical they come at a pretty hefty CFL criteria cost.  If it is indeed dead simple to implement at cell boundaries it'd be really nice to start there and then if someone cares to do the fancier algorithm that could still be done.  

Just for context I almost always leave `hFac=1` in the MITgcm for flow over bathymetry problems because my vertical grid cells tend to be small compared to the topography, and w tends to be large, so my vertical CFL criteria is usually the limiting factor.  And, as pointed out above, the ocean is rough, and any effects from a perfectly smooth Gaussian bump versus a slightly stair-steppy  one had better be negligible if what I am hoping to say has any real-world importance.

**Reply by glwagner** [reply]
I was more thinking of LES with resolutions of a fraction of a meter to a few meters (what we primarily use it for now). You still might be right that accurate grid information isn't too important for many or most problems even at that scale.

There would be interest in using Oceananigans for DNS in complex geometries if it were capable of that, I think. But you're right its not our main focus. I buy your argument that stair-steppiness isn't important for non-LES problems for sure.

I guess my reading of the literature is that a generic immersed boundary method may not much more difficult than a grid coincident one, provided the boundary is not moving. For such a problem we don't *think* we need to modify the pressure solver (or if we do, the changes will hopefully be minor; eg using subiterations of the immersed boundary correction + pressure solve). In that case the difference between a grid-coincident method and a more general method "just" requires a bit of geometry.

MITgcm does a direct solve of the interior pressure poisson problem, which may be a source of complexity in implementing "shaved cells", etc? I'm not sure. There is some dissonance between a description of the MITgcm algorithm and engineering papers describing immersed boundary implementations that I don't fully understand.

Over on #1036 there's some discussion about @whitleyv's immersed boundary effort. @whitleyv and some collaborators were able to get a grid-coincident method spun up pretty quickly (approx a day of typing I think --- there's some results on #1036, @ali-ramadhan also has a few more involving a rising thermal bubble).

**Comment by navidcy** [comment]
> I was looking over the documentation for oceananigans. It is looking very exciting, though I don't imagine I can seriously use it until MPI and immersed boundaries work.
> 
> I found the discussion of the "Powerful algorithm" in https://clima.github.io/OceananigansDocumentation/stable/generated/kelvin_helmholtz_instability/ a bit more obscure than perhaps necessary. It seems that you are successively filtering the solutions so that you emphasize the fastest growing mode, but keep the amplitude of the velocities in a regime that will let the model stay linear. What I was missing is a clearer motivation of "why". Can you include a couple of sentences about what the end goal of this would be, and why it is not achievable just running a free simulation? I _think_ its something like:
> 
> > The goal is to find the fastest growing linear mode, and have an idea of its growth rate to compare with theory.  This is difficult to extract empirically from a free simulation because the non-linearities quickly spread variance from the fastest growing mode to other scales.
> 
> but I really had to think to dredge that up, and I'm not sure if I am even correct.

That’s a good point! Thanks. 
I’m offline at the moment but I’ll add a remark along those lines in the example when I’m back in civilization in a few days.

**Comment by francispoulin** [comment]
@jklymak raises some good points.  In terms of the power method, I agree tha tit can be made clearer.  I think the basic idea is that we want to get the growth rate and spatial structure in the context of the linear theory.  I think this should be equivalnet to just solving the linearized eqations.  I wonder if making this connection might help?

I am also keen to help with the MPI issue.  The plan is to start testing with ShallowWaterModel since there is no elliptic problem to solve and hence a bit easier.  This is for when people have time of course.
